### ISSUE TITLE
web.context-root is needed if the URL can not be rewritten.

### ISSUE THREAD

**Issue by chi-bd** [TwinProduction/gatus#88]
At a1679ddc5e6c36e58fcfc75ece86bbe557e27709, `web.context-root` is deleted.
But this option is needed in the environment which the URL can not be rewritten, for example Amazon ALB ingress controller.
Would you please revive it?

**Comment by TwinProduction** [comment]
As mentioned in v2.0.0's release notes:
> Removed web.context-root (Vue.js' router doesn't easily support context roots. May be re-implemented in the future if enough people ask for it)

The main reason for removing it is due to my inability to make a dynamic context root work with Vue.js' router (without ugly hacks) despite several attempts. 

If you want to have a go at it, feel free, but unless a few more people request this, I won't personally try to re-implement it.

That being said, for your specific issue, most people create a separate CNAME (e.g. `status.example.com`) for their status page. 
Is there something preventing you from doing this?

**Comment by KarlMW** [comment]
What about this?  https://stackoverflow.com/questions/43879418/deploy-vuejs-app-in-a-subdirectory
It seems that it can be done with an edit of `web/app/vue.config.js` like this:
```
module.exports = {
  ...
  publicPath: "./"
  ...
};
```
(bear in mind that I know nothing of what I write - I just googled the problem - hopefully not wasting your time here)

In terms of the usefulness of being able to host content in a sub-directory - it's one of the first things I look for in a service I want to privately self-host (ie. purely for my own use).  Being in a subdirectory means that its existence is not visible, whereas the existence of subdomains is publicly displayed in DNS.

**Comment by KarlMW** [comment]
I hope that's not one of the "ugly" hacks you refer to  :-)

**Comment by TwinProduction** [comment]
@KarlMW Hahaha no, if it was just that, it couldn't even be called a hack.

In fact, this would work fine if I weren't using vue router for the service detail page.

Long story short, one of `publicPath`'s limitations is that it cannot a relative path cannot be used in conjunction with `history.pushState`, which is unfortunately used by Vue.js' router. I could fix this by using hash history (e.g. `http://localhost:8080/#/services/core_twinnation-sitemap`), but I just can't bring myself to like that style of routing.

[reference](https://cli.vuejs.org/config/#publicpath):
> ### Limitations of relative publicPath
>
> Relative publicPath has some limitations and should be avoided when:
> - You are using HTML5 history.pushState routing;
> - You are using the pages option to build a multi-paged app.

I tried several different methods, but none worked in a clean way.

I didn't want to specify what I meant by "ugly" hacks, because they're really ugly, but the one way I found that works would be to read the generated static script (`https://github.com/TwinProduction/gatus/blob/master/web/static/js/app.js`) on application start, replace the value of a constant by the defined context root and overwrite the default script with the new script that has the appropriate context root. Coupling that with the `base` parameter of VueRouter (https://router.vuejs.org/api/#base) would be sufficient, but well, it's hacky.

----

By the way, it is not possible to do a reverse lookup to retrieve the CNAMEs of a domain (unless you have access to the zone files).

For instance, if I create the CNAME `potato.twinnation.org`, it's _basically_ impossible for anybody to know unless I tell them, or unless they bruteforce every single strings and find it _(or they just happen to try it - but don't bother, I don't have that cname 🤣)_.

This is known as **denial of existence**. 

Even if a malicious user had access to the zone files, you could still hide it from them by using a wildcard CNAME domain record, and by pointing that to a reverse proxy which would have the necessary configuration telling it to route requests with a given host to the appropriate application, you'd be the only person privy to the existence of that application.

That being said, this is still security by obscurity, which is IMO not enough if you don't want prying eyes to have access to your application, which is why you can also configure basic authentication on top of that: https://github.com/TwinProduction/gatus#basic-authentication

**Comment by KarlMW** [comment]
Thanks for the detailed reply.  You've made gatus very easy for people to build in docker themselves - thanks for that, too  :-)
I tried setting `publicPath: '/gatus/'` in `web/app/vue.config.js` and then rebuilt it, hoping that a static root would work.  Sadly it failed - it was still looking for its assets under `/`

w.r.t. CNAME reverse lookups, I'm not sure how it's done, but my never-used-publicly CNAMEs are findable on the web, eg. via https://www.nmmapper.com/sys/tools/subdomainfinder/

Whether it's some sort of active probing, or whether they just monitor certificate transparency logs, I don't know.  I generally use client certificates (handled by traefik) for my private domains (as well as passwords), so I'm reasonably confident that no-one can get in, but I like a bit of obscurity as well  :-)

**Comment by MiddleMan5** [comment]
Allowing us to configure the full external path would also be acceptable. You may still need to modify a constant on the client side (our backends usually write the configured value to <base href=PATH /> which our frontend's read into a variable in our path building utils on load)

IE:
`external-root: https://health.my-domain.com/region-1/`

All relative links accessed by the client would be replaced by the full external root path, although I'm not sure how this plays with pushState.

**Comment by TwiN** [comment]
@MiddleMan5 If you or anybody else can get it to work with the front-end, feel free to do so. I'm completely fine with using a different approach for routing -- if `pushState` has to go, then so be it. 

I do want to rewrite a good part of the UI at some point and make it a little less minimalist, but I don't know when that will happen.

**Comment by michaelkrieger** [comment]
> w.r.t. CNAME reverse lookups, I'm not sure how it's done, but my never-used-publicly CNAMEs are findable on the web, eg. via https://www.nmmapper.com/sys/tools/subdomainfinder/

They get this data from SSL Transparency.  Looking at it from a number of domains I control, it includes only those that exist in SSL Transparency records (even a typo that existed for 1 minute) that show up there.  Unless you have an misconfigured (or old pre [RFC8482](https://www.rfc-editor.org/rfc/rfc8482) server that still answers ANY queries, this is the only place to get the values.

I'm sure the big names (8.8.8.8 and 1.1.1.1 or major ISPs) could probably assemble it from DNS query logging though.  It should be assumed that it's out there (and hence have basic auth or something like authelia in front of it). I agree that subdomains are a better place for a lot of services, but this same problem exists with many VUE apps.

**Comment by ansilh** [comment]
Got stuck in this issue when I configured Gatus behind a Google's LB backed by IAP authentication.
we use a common FQDN and uses route based redicretion via URL map rewrite, but the Gatus web app always respond with default root ("/") instead of the called base URL ("https://common.ingress.dom/<custom_dir>/")
Now the only option for left is to create an Ingress rule to route the traffic via a new DNS entry and I've to duplicate the GCP ingress configuration only for Gatus :( 

Would be great if we can call this out in the README to avoid surprise at the end.

**Comment by TwiN** [comment]
@ansilh Could you make a PR for this? 
A section in the FAQ should suffice

**Comment by akozlins** [comment]
after some tries, here is what worked for me:

- rebuild gatus fronend (`npm --prefix web/app run build`) with following changes
```
# set publicPath in vue.config.js
ARG VUE_PUBLIC_PATH
RUN sed -i \
    -e "s|publicPath: .*|publicPath: '${VUE_PUBLIC_PATH}'|" \
    web/app/vue.config.js

# use relative path '.' for SERVER_URL in main.js
RUN sed -i \
    -e "s|export const SERVER_URL = .*|export const SERVER_URL = '.'|" \
    web/app/src/main.js

# use relative path '.' for resources in index.html
RUN sed -i \
    -e 's|="/|="./|g' \
    web/app/public/index.html
```

- and reverse proxy (only relevant lines from docker-compose)
```
build: { dockerfile: "Dockerfile", args: [ "VUE_PUBLIC_PATH=/gatus", ] }
labels:
- "traefik.http.middlewares.gatus.stripprefix.prefixes=/gatus"
- "traefik.http.routers.gatus.middlewares=gatus"
- "traefik.http.routers.gatus.rule=PathPrefix(`/gatus/`)"
```

**Comment by rokiden** [comment]
I'm homelab enthusiast with free domain from DDNS provider, subdomain option unavailable for me. Subpath would be very useful, otherwise I can use Gatus only through SSH tunnel :(

**Comment by lpil** [comment]
I'm in the same situation as @rokiden here!

Does Gatus UI have multiple routes? I couldn't work out how to get any other routes to the only paths that would need to change would be the CSS and JS, both of which could be handled by removing the `/` from the start of the path to them if there is only one route.

**Comment by serGrey** [comment]
> after some tries, here is what worked for me:
> 
> * rebuild gatus fronend (`npm --prefix web/app run build`) with following changes
> 
> ```
> # set publicPath in vue.config.js
> ARG VUE_PUBLIC_PATH
> RUN sed -i \
>     -e "s|publicPath: .*|publicPath: '${VUE_PUBLIC_PATH}'|" \
>     web/app/vue.config.js
> 
> # use relative path '.' for SERVER_URL in main.js
> RUN sed -i \
>     -e "s|export const SERVER_URL = .*|export const SERVER_URL = '.'|" \
>     web/app/src/main.js
> 
> # use relative path '.' for resources in index.html
> RUN sed -i \
>     -e 's|="/|="./|g' \
>     web/app/public/index.html
> ```
> 
> * and reverse proxy (only relevant lines from docker-compose)
> 
> ```
> build: { dockerfile: "Dockerfile", args: [ "VUE_PUBLIC_PATH=/gatus", ] }
> labels:
> - "traefik.http.middlewares.gatus.stripprefix.prefixes=/gatus"
> - "traefik.http.routers.gatus.middlewares=gatus"
> - "traefik.http.routers.gatus.rule=PathPrefix(`/gatus/`)"
> ```

could you please help me with it

**Comment by simonhammes** [comment]
I just opened a PR that tries to fix this: #908

**Comment by semik** [comment]
I'm testing Gatus as a status page for our product CZERTAINLY. It operates in Kubernetes and comprises several services which we would like to monitor. We have already integrated KeyCloak for the SAML/OIDC proxy and RabbitMQ as a messaging bus. Keycloak resides on $URL/kc, RabbitMQ on $URL/mq, and we would like to integrate Gatus the same way - to reside on $URL/status.

I like @simonhammes solution, which made this possible.

**Comment by digitalray** [comment]
> > after some tries, here is what worked for me:
> > 
> > * rebuild gatus fronend (`npm --prefix web/app run build`) with following changes
> > 
> > ```
> > # set publicPath in vue.config.js
> > ARG VUE_PUBLIC_PATH
> > RUN sed -i \
> >     -e "s|publicPath: .*|publicPath: '${VUE_PUBLIC_PATH}'|" \
> >     web/app/vue.config.js
> > 
> > # use relative path '.' for SERVER_URL in main.js
> > RUN sed -i \
> >     -e "s|export const SERVER_URL = .*|export const SERVER_URL = '.'|" \
> >     web/app/src/main.js
> > 
> > # use relative path '.' for resources in index.html
> > RUN sed -i \
> >     -e 's|="/|="./|g' \
> >     web/app/public/index.html
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > 
> > * and reverse proxy (only relevant lines from docker-compose)
> > 
> > ```
> > build: { dockerfile: "Dockerfile", args: [ "VUE_PUBLIC_PATH=/gatus", ] }
> > labels:
> > - "traefik.http.middlewares.gatus.stripprefix.prefixes=/gatus"
> > - "traefik.http.routers.gatus.middlewares=gatus"
> > - "traefik.http.routers.gatus.rule=PathPrefix(`/gatus/`)"
> > ```
> 
> could you please help me with it

I am not following this. Here is what I did:

export VUE_PUBLIC_PATH=/healthcheck

```shell
# set publicPath in vue.config.js
sed -i -e "s|publicPath: .*|publicPath: '${VUE_PUBLIC_PATH}'|" web/app/vue.config.js

# use relative path '.' for SERVER_URL in main.js
sed -i -e "s|export const SERVER_URL = .*|export const SERVER_URL = '.'|" web/app/src/main.js

# use relative path '.' for resources in index.html
sed -i  -e 's|="/|="./|g' web/app/public/index.html

# Install tooling
npm install -g @vue/cli
npm i @vue/cli-service
npm i '@vue/cli-plugin-babel'
npm i '@vue/cli-plugin-eslint'
npm i 'vue-template-compiler'

# install npm dependencies
npm --prefix web/app install
npm --prefix web/app run build

# build docker image
docker build --platform=linux/amd64 . -t ${FULL_IMAGE_NAME}
```
Still getting 404 on /healthcheck. I am using the dockerfile that comes with the repo. I think this is where the issue is. After you build the front-end, how do you ensure that build ends up in the image?

**Comment by akozlins** [comment]
> ```shell
> # curl request
> ❯ curl -i localhost:8080/healthcheck
> HTTP/1.1 404 Not Found
> Date: Tue, 26 Nov 2024 06:03:45 GMT
> Content-Type: text/plain; charset=utf-8
> Content-Length: 23
> 
> Cannot GET /healthcheck
> ```
> 
> Still getting 404 on /healthcheck. I am using the dockerfile that comes with the repo. I think this is where the issue is. After you build the front-end, how do you ensure that build ends up in the image?

note that app is still served on `/`

the `publicPath` makes some generated urls on the page use `VUE_PUBLIC_PATH` as root for links,
so you still need to do reverse proxy from `/healthcheck` to `/` (e.g. via nginx/caddy/traefik)

for image i just use custom Dockerfile (e.g. <https://github.com/akozlins/server/blob/master/gatus/Dockerfile>)

**Comment by EliranTurgeman** [comment]
Any update on this case?

**Comment by roughnecks** [comment]
I would also like to see this happen!

**Comment by biomassboilers** [comment]
This is a blocking issue for us to Gatus, too. It's a very good tool, but can't dedicate a server to it and need to run it under a subdomain at `http://servername.local/gatus`...

**Comment by TwiN** [comment]
There's already a PR up, but one of my comments on it hasn't been addressed yet. Given how long it's been, I'm also open to people creating a separate PR to make an attempt at fixing it. My only requirement is that it is not a breaking change.

This is not a feature I will personally use, and I focus on features I personally want to use during my free time, so I hope you understand this is why it still hasn't been addressed.

Despite that, I personally already gave it a shot a few years ago because I knew a lot of people wanted this (see https://github.com/TwiN/gatus/issues/88#issuecomment-787583120), alas, I wasn't able to get it working. I imagine the "real" fix here would be to change the router used by the frontend, though given how long it has been, perhaps it has already been fixed upstream.

**TL;DR**: I will gladly look over any PR made to implement it, but I will not implement it myself. I already have a roadmap of features I want to prioritize implementing, and this isn't one of them, at least for the short term.

**Comment by TwiN** [comment]
There's already a PR up, but one of my comments on it hasn't been addressed yet. Given how long it's been, I'm also open to people creating a separate PR to make an attempt at fixing it. My only requirement is that it is not a breaking change for existing users.

This is not a feature I will personally use, and I focus on features I personally want to use during my free time, so I hope you understand this is why it still hasn't been addressed.

Despite that, I personally already gave it a shot a few years ago because I knew a lot of people wanted this (see https://github.com/TwiN/gatus/issues/88#issuecomment-787583120), alas, I wasn't able to get it working. I imagine the "real" fix here would be to change the router used by the frontend, though given how long it has been, perhaps it has already been fixed upstream.

**TL;DR**: I will gladly look over any PR made to implement it, but I will not implement it myself. I already have a roadmap of features I want to prioritize implementing, and this isn't one of them, at least for the short term.

**Comment by biomassboilers** [comment]
I've been using a custom build https://github.com/simonhammes/gatus/commit/b8d699fa5368872b1a1f3f6f0011077483b45ea0 (see PR https://github.com/TwiN/gatus/pull/908), made with with https://github.com/thongtech/go-legacy-win7/releases/tag/v1.25.0-2

**Comment by DarkWiiPlayer** [comment]
> That being said, for your specific issue, most people create a separate CNAME (e.g. status.example.com) for their status page.

For someone like me who hosts stuff as a hobby for me and my friends to use, this may be technically viable, but is just too inconvenient in practice. It's not that much more effort for a single application, but if every thing I want to host needs a new subdomain, that'd just be too much effort to be worth it.

Having nginx as the one and only place where different applications need to be set up and managed makes things a lot easier.

**Comment by oschrenk** [comment]
Would love to see this moving forward.

**Comment by shornabho** [comment]
Have taken the liberty to solve this for my own use case, and it's pushed to #1726. 

My patched _unofficial_ pre-release version is also available to use from `ghcr.io/shornabho/gatus:5.37.0-basepath.1`, if anyone wants to use. Feel free to come back with any suggestions on the PR or any issues you find.
